dataplane/linux/vxlan_mgr.go

```diff
@@ -0,0 +1,307 @@
+	case *proto.VXLANTunnelEndpointUpdate:
+		logrus.WithField("msg", msg).Warnf("VXLAN data plane received VTEP update")
+		if msg.Node == m.hostname {
+			m.myVTEP = msg
+		} else {
+			m.pendingVTEPUpdates[msg.Node] = msg
+		}
+	case *proto.VXLANTunnelEndpointRemove:
+		logrus.WithField("msg", msg).Warnf("VXLAN data plane received VTEP remove")
+		if msg.Node == m.hostname {
+			m.myVTEP = nil
+		}
+	}
+}
+
+func (m *vxlanManager) CompleteDeferredWork() error {
+	// Update state based on pending updates.
+	l2routes := []routetable.L2Target{}
+	for _, u := range m.pendingVTEPUpdates {
+		mac, err := net.ParseMAC(u.Mac)
+		if err != nil {
+			return err
+		}
+		l2routes = append(l2routes, routetable.L2Target{
+			IP:      ip.FromString(u.NodePublicIp),
+			VTEPMAC: mac,
+			GW:      ip.FromString(u.Ipv4Addr),
+		})
+		m.sentVTEPs[u.Node] = u
+	}
+	if len(l2routes) > 0 {
+		logrus.WithField("l2routes", l2routes).Warnf("VXLAN manager sending l2 updates")
+		m.routeTable.SetL2Routes("vxlan.calico", l2routes)
+		m.pendingVTEPUpdates = map[string]*proto.VXLANTunnelEndpointUpdate{}
+	}
+
+	// Iterate through all of our routes and send them through to the route table.
+	routes := []routetable.Target{}
+	for _, r := range m.pendingRouteUpdates {
+		logCtx := logrus.WithField("route", r)
+		cidr, err := ip.CIDRFromString(r.Dst)
+		if err != nil {
+			return err
+		}
+
+		// Extract the gateway addr for this route based
+		// on its remote VTEP.
+		vtep, ok := m.sentVTEPs[r.Node]
+		if !ok {
+			logCtx.Warnf("Dataplane has route with no corresponding VTEP")
+			continue
+		}
+
+		routes = append(routes, routetable.Target{
+			Type: routetable.TargetTypeVXLAN,
+			CIDR: cidr,
+			GW:   ip.FromString(vtep.Ipv4Addr),
+		})
+	}
+
+	if len(routes) > 0 {
+		logrus.WithField("routes", routes).Warnf("VXLAN manager completing deferred work")
+		m.routeTable.SetRoutes("vxlan.calico", routes)
+		m.pendingRouteUpdates = map[string]*proto.RouteUpdate{}
+	}
+	return nil
+}
+
+// KeepVXLANDeviceInSync is a goroutine that configures the VXLAN tunnel device, then periodically
+// checks that it is still correctly configured.
+func (m *vxlanManager) KeepVXLANDeviceInSync() {
+	logrus.Info("VXLAN tunnel device thread started.")
+	for {
+		if m.myVTEP == nil {
+			logrus.Warn("Missing local VTEP information, retrying...")
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		err := m.configureVXLANDevice()
+		if err != nil {
+			logrus.WithError(err).Warn("Failed configure VXLAN tunnel device, retrying...")
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		time.Sleep(10 * time.Second)
+	}
+}
+
+// configureVXLANDevice ensures the VXLAN tunnel device is up and configures correctly.
+func (m *vxlanManager) configureVXLANDevice() error {
+	logCxt := logrus.WithFields(logrus.Fields{})
+	logCxt.Debug("Configuring VXLAN tunnel")
+	link, err := netlink.LinkByName("vxlan.calico")
+	if err != nil {
+		parent, err := netlink.LinkByName("ens4")
```

Comment:
yeah, I think we'll need something. Was wondering if we should just find the interface that has the node's IP address on it.

We could look at `myVTEP` and see what the parent IP is, then find that interface and grab it here. That minimizes the configuration required since it's basically derived from the IP autodetection logic in calico/node which is already fairly flexible (or whatever logic is already setting the node IP in environments where calico/node isn't used).